### PR TITLE
[PowerPC] Alignment of toc-data symbol should not be increased during optimizations

### PR DESCRIPTION
**Pull request by bzEq**
Currently, the alignment of toc-data symbol might be changed during instcombine
```
IC: Visiting:   %global = alloca %struct.widget, align 8                                                                                         
Found alloca equal to global:   %global = alloca %struct.widget, align 8                                                                         
  memcpy =   call void @llvm.memcpy.p0.p0.i64(ptr nonnull align 1 %global, ptr align 1 @global, i64 3, i1 false)
```
The `alloca` is created with `PrefAlign` which is 8 and after IC, the alignment of `@global` is enforced into `8`, same as the `alloca`. This is not expected, since toc-data symbol has the same alignment as toc entry and should not be increased during optimizations.

### CODE REVIEW

File: llvm/lib/IR/Globals.cpp [hunk 1]
```diff
@@ -335,6 +335,14 @@ bool GlobalObject::canIncreaseAlignment() const {
   if (isELF && !isDSOLocal())
     return false;
 
+  bool isXCOFF =
+      (!Parent || Triple(Parent->getTargetTriple()).isOSBinFormatXCOFF());
+  if (isXCOFF)
+    if (const GlobalVariable *GV = dyn_cast<GlobalVariable>(this))
+      // GV with toc-data attribute is put in the region same as toc entry.
```
Comment (mandlebug) [inline llvm/lib/IR/Globals.cpp:342]:
```suggestion
      // GV with toc-data attribute is defined in a TOC entry which 
      // has a fixed alignment and cannot be arbitrarily increased.
```

File: llvm/test/CodeGen/PowerPC/tocdata-firm-alignment.ll
Comment (mandlebug) [inline llvm/test/CodeGen/PowerPC/tocdata-firm-alignment.ll:1]: Would it be better suited to place the test in `Transforms/InferAlignment/` instead?

**Review by mandlebug — APPROVED** [review]

File: llvm/test/CodeGen/PowerPC/tocdata-firm-alignment.ll
Comment (bzEq) [reply llvm/test/CodeGen/PowerPC/tocdata-firm-alignment.ll:1]: That looks like a place for `InferAlignmentPass`.

**Review by syzaara — APPROVED** [review]

**Comment by MaskRay** [comment]
>  This is not expected, since toc-data symbol has the same alignment as toc entry and should not be increased during optimizations.

Why does AIX have this requirement? ppc64 ELFv2 ABI can increase the alignment without any issue.

**Review by efriedma-quic — COMMENTED** [review]
From the perspective of the IR, this is the correct way to implement this restriction.  (toc-data should be documented in LangRef, though.)  I won't try to dive into the XCOFF side of this.

**Comment by bzEq** [comment]
> Why does AIX have this requirement? ppc64 ELFv2 ABI can increase the alignment without any issue.

I tend to believe it's implementation's limitation, see https://github.com/llvm/llvm-project/blob/f2441b0297501610b59527f93e3174c814d63941/llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp#L2727.

**Comment by chenzheng1030** [comment]
> Why does AIX have this requirement? ppc64 ELFv2 ABI can increase the alignment without any issue.

IIUC, AIX has limited TOC region especially for small code model and TOC region is created by linker(i.e., compiler can not fully control the layout of the TOC entries), so to avoid wasting of TOC entries, guarding alignment of the TOC data symbols with normal toc entries is necessary.

```
int a;
int foo()
{
  return a;
}

int b;
int main() {
  return b + foo();
}
```

If a is a toc-data symbol and we change its alignment to 16 in the assembly ( 2^16 is the size of the TOC region for small code model), compiling the modified assembly will make AIX linker emit TOC overflow error.

**Comment by MaskRay** [comment]
> > Why does AIX have this requirement? ppc64 ELFv2 ABI can increase the alignment without any issue.
> 
> IIUC, AIX has limited TOC region especially for small code model and TOC region is created by linker(i.e., compiler can not fully control the layout of the TOC entries), so to avoid wasting of TOC entries, guarding alignment of the TOC data symbols be same with normal toc entries is necessary.
> 
> ```
> int a;
> int foo()
> {
>   return a;
> }
> 
> int b;
> int main() {
>   return b + foo();
> }
> ```
> 
> If a is a toc-data symbol and we change its alignment to 16 in the assembly ( 2^16 is the size of the TOC region for small code model), compiling the modified assembly will make AIX linker emit TOC overflow error.

Thanks for the example. I have also tried

```
clang -O2 -c a.c --target=powerpc64 -mabi=elfv{1,2} -mcmodel={small,medium}
```

AIX has the limited R_TOC relocation, similar to R_PPC64_TOC16_DS emitted by ELFv1 ELFv2 with -mcmodel=small.
I think nearly nobody uses -mcmodel=small for ELF, so having this exception only for AIX makes sense.

File: llvm/lib/IR/Globals.cpp [hunk 2]
```diff
@@ -335,6 +335,14 @@ bool GlobalObject::canIncreaseAlignment() const {
   if (isELF && !isDSOLocal())
     return false;
 
+  bool isXCOFF =
+      (!Parent || Triple(Parent->getTargetTriple()).isOSBinFormatXCOFF());
+  if (isXCOFF)
+    if (const GlobalVariable *GV = dyn_cast<GlobalVariable>(this))
+      // GV with toc-data attribute is defined in a TOC entry which has a fixed
```
Comment (MaskRay) [inline llvm/lib/IR/Globals.cpp:342]: This comment can be lifted before `isXCOFF`. It seems not correct.

The alignment is not increased to mitigate TOC size overflow.

**Review by MaskRay — APPROVED** [review]

**Comment by bzEq** [comment]
> toc-data should be documented in LangRef, though

Posted https://github.com/llvm/llvm-project/pull/94828.

**Comment by bzEq** [comment]
I'll leave doc change also in the same PR.

**Comment by bzEq** [comment]
Ping @efriedma-quic , could you please review the doc change?

File: llvm/lib/IR/Globals.cpp [hunk 3]
```diff
@@ -335,6 +335,15 @@ bool GlobalObject::canIncreaseAlignment() const {
...
     return false;
 
+  // GV with toc-data attribute is defined in a TOC entry. To mitigate TOC
+  // overflow, the alignment of such symbol should not be increased. Otherwise,
+  // padding is needed thus more TOC entries are wasted.
+  bool isXCOFF =
+      (!Parent || Triple(Parent->getTargetTriple()).isOSBinFormatXCOFF());
+  if (isXCOFF)
+    if (const GlobalVariable *GV = dyn_cast<GlobalVariable>(this))
+      return !GV->hasAttribute("toc-data");
```
Comment (efriedma-quic) [inline llvm/lib/IR/Globals.cpp:345]:
```suggestion
      if (GV->hasAttribute("toc-data"))
        return false;
```

**Review by efriedma-quic — APPROVED** [review]
New documentation looks fine.  One minor suggestion for the code; otherwise LGTM

**Pull request merged**